health/health.go

```diff
@@ -0,0 +1,89 @@
+package health
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/gofiber/fiber/v2"
+	"github.com/sirupsen/logrus"
+	"go.uber.org/fx"
+)
+
+type status struct {
+	Healthy bool
+	Reason  *string
+
+	mux sync.Mutex
+}
+
+var serviceStatus = &status{
+	Healthy: true,
+}
+
+var Module = fx.Invoke(
+	useHealthEndpoint,
+	useHealthLogger,
+)
+
+func useHealthEndpoint(app *fiber.App) {
+	app.Get("/health", func(c *fiber.Ctx) error {
+		if !serviceStatus.Healthy {
+			return c.Status(503).SendString(
+				fmt.Sprintf("Service unavailable: %s", *serviceStatus.Reason),
+			)
+		}
+		return c.Status(200).SendString("Healthy")
+	})
+}
+
+func useHealthLogger(lifecycle fx.Lifecycle, logrus *logrus.Logger) {
+	ctx, cancel := context.WithCancel(context.Background())
+	healthy := true
+
+	lifecycle.Append(fx.Hook{
+		OnStop: func(ctx context.Context) error {
+			cancel()
+			return nil
+		},
+	})
+
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				if healthy != serviceStatus.Healthy {
+					if healthy {
+						logrus.Info("Healthy")
+					} else {
+						logrus.WithField("reason", *serviceStatus.Reason).
+							Warn("Service unavailable")
+					}
+				}
+				healthy = serviceStatus.Healthy
+			}
+		}
+	}()
+}
+
+func IsHealthy() bool {
+	return serviceStatus.Healthy
+}
+
+func StatusHealthy() {
```

Comment:
As different parts (components) of the app can say that they are healthy or not:

What about adding a `component string` parameter to `StatusHealthy` and `StatusNotHealthy`.
We could then have a `map[string]status` that stores the healthy state of all components of the app. As soon as one of those components is unhealthy the whole app is unhealthy. This eliminates the risk of a component overwriting the healthy state of an other component.

Alternative: just remove the `StatusHealthy` function. Downside: a service will be killed as soon as some component gets unhealthy for a short period and there is no way to recover.

Addition: can we please rename the functions so it becomes clearer that they are updating the healthy state, e.g. `SetStatusHealthy`. Without having a look at return types they sound like getter functions.

Reply:
Ok, I will rework that